### PR TITLE
Issue #125:Adding enricher for Wcf instrumentation

### PR DESCRIPTION
**Pull request by epaulsen**

### CODE REVIEW

File: src/OpenTelemetry.Contrib.Instrumentation.Wcf/README.md [hunk 1]
```diff
@@ -20,6 +20,24 @@ using var openTelemetry = Sdk.CreateTracerProviderBuilder()
...
+Instrumentation can be configured via options overload for `AddWcfInstrumentation` method:
+
+```csharp
+using var openTelemetry = Sdk.CreateTracerProviderBuilder()
+    .AddWcfInstrumentation(options => 
+    {
+        options.SuppressDownstreamInstrumentation = false;  
+        options.Enrich = (activity,eventName,message) =>    // -> Enable enriching an activity after it is created.
+        {
+            if(eventName == "AfterReceiveRequest")  // For event names, please refer to string contstants in WcfEventNames class.
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Contrib.Instrumentation.Wcf/README.md:32]: Might be nice to use `WcfEventNames` in the example so people know it is available.

File: src/OpenTelemetry.Contrib.Instrumentation.Wcf/TelemetryClientMessageInspector.cs [hunk 2]
```diff
@@ -111,6 +110,8 @@ public object BeforeSendRequest(ref Message request, IClientChannel channel)
                     {
                         activity.SetTag(WcfInstrumentationConstants.SoapViaTag, request.Properties.Via.ToString());
                     }
+
+                    WcfInstrumentationActivitySource.Options.Enrich?.Invoke(activity, WcfEventNames.BeforeSendRequest, request);
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Contrib.Instrumentation.Wcf/TelemetryClientMessageInspector.cs:114]: Would you mind wrapping these calls in try/catch? Example here:

https://github.com/open-telemetry/opentelemetry-dotnet/blob/5ddf9a486e755c53ab73debf87286a934fcbbb51/src/OpenTelemetry.Instrumentation.AspNet/Implementation/HttpInListener.cs#L152-L159

File: src/OpenTelemetry.Contrib.Instrumentation.Wcf/WcfEventNames.cs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+namespace OpenTelemetry.Contrib.Instrumentation.Wcf
+{
+    /// <summary>
+    /// Constants used for event names when enriching an activity.
+    /// </summary>
+    public class WcfEventNames
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Contrib.Instrumentation.Wcf/WcfEventNames.cs:22]: The name is a bit generic, what if we changed to something like `WcfEnrichEventNames`?

File: src/OpenTelemetry.Contrib.Instrumentation.Wcf/README.md (hunk 1 above)
Comment (epaulsen) [reply src/OpenTelemetry.Contrib.Instrumentation.Wcf/README.md]: Absolutely.  Done.

File: src/OpenTelemetry.Contrib.Instrumentation.Wcf/TelemetryClientMessageInspector.cs (hunk 2 above)
Comment (epaulsen) [reply src/OpenTelemetry.Contrib.Instrumentation.Wcf/TelemetryClientMessageInspector.cs]: Done.  Also added same exception logging to `TelementryDispatchMessageInspector`

File: src/OpenTelemetry.Contrib.Instrumentation.Wcf/WcfEventNames.cs (hunk 3 above)
Comment (epaulsen) [reply src/OpenTelemetry.Contrib.Instrumentation.Wcf/WcfEventNames.cs]: Renamed 👍

**Comment by epaulsen** [comment]
@CodeBlanch , thanks for the review.  
I've updated my pullrequest based on your comments, do you mind having another look?

**Review by CodeBlanch — APPROVED** [review]
Thanks @epaulsen this LGTM! I'm going to leave it open for a day or two just in case anyone else wants to also take a look at it.

File: src/OpenTelemetry.Contrib.Instrumentation.Wcf/README.md [hunk 4]
```diff
@@ -20,6 +20,29 @@ using var openTelemetry = Sdk.CreateTracerProviderBuilder()
...
+        options.SuppressDownstreamInstrumentation = false;  
+
+        // Enable enriching an activity after it is created.
+        options.Enrich = (activity,eventName,message) =>    
+        {
+            // For event names, please refer to string constants in 
+            // WcfEnrichEventNames class.
+            if(eventName == WcfEnrichEventNames.AfterReceiveRequest)  
+            {
+                activity.AddTag("companyname.customtag", Guid.NewGuid());
```
Comment (cijothomas) [inline src/OpenTelemetry.Contrib.Instrumentation.Wcf/README.md:39]:
```suggestion
                activity.SetTag("companyname.customtag", Guid.NewGuid());
```

Comment (cijothomas) [reply src/OpenTelemetry.Contrib.Instrumentation.Wcf/README.md:39]: Guid is not a supported type for span/activity attributes/tags.

File: src/OpenTelemetry.Contrib.Instrumentation.Wcf/README.md [hunk 5]
```diff
@@ -20,6 +20,29 @@ using var openTelemetry = Sdk.CreateTracerProviderBuilder()
...
+`AddWcfInstrumentation` method:
+
+```csharp
+using var openTelemetry = Sdk.CreateTracerProviderBuilder()
+    .AddWcfInstrumentation(options => 
+    {
+        options.SuppressDownstreamInstrumentation = false;  
+
+        // Enable enriching an activity after it is created.
+        options.Enrich = (activity,eventName,message) =>
```
Comment (cijothomas) [inline src/OpenTelemetry.Contrib.Instrumentation.Wcf/README.md:33]: "message" is not used in this example anywhere. I'd suggest to have it used in the example. If it were not for the message, then ActivityProcessor is the recommendation for enriching any activity.

see example: https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/src/OpenTelemetry.Instrumentation.AspNetCore#enrich

File: src/OpenTelemetry.Contrib.Instrumentation.Wcf/WcfInstrumentationOptions.cs [hunk 6]
```diff
@@ -34,6 +35,18 @@ public class WcfInstrumentationOptions
...
+        /// Gets or sets an action to enrich an Activity.
+        /// </summary>
+        /// <remarks>
+        /// <para><see cref="Activity"/>: the activity being enriched.</para>
+        /// <para>string: the name of the event. Will be one of the constants in <see cref="WcfEnrichEventNames"/>.
+        /// </para>
+        /// <para>object: the raw <see cref="Message"/> from which additional information can be extracted to enrich the activity.
+        /// </para>
+        /// </remarks>
+        public Action<Activity, string, object> Enrich { get; set; }
```
Comment (cijothomas) [inline src/OpenTelemetry.Contrib.Instrumentation.Wcf/WcfInstrumentationOptions.cs:48]: it';d be good to add an entry in changelog.md as well for this new feature

Comment (epaulsen) [reply src/OpenTelemetry.Contrib.Instrumentation.Wcf/WcfInstrumentationOptions.cs:48]: Done.

File: src/OpenTelemetry.Contrib.Instrumentation.Wcf/README.md (hunk 4 above)
Comment (epaulsen) [reply src/OpenTelemetry.Contrib.Instrumentation.Wcf/README.md]: Code example changed, see comment below.

File: src/OpenTelemetry.Contrib.Instrumentation.Wcf/README.md (hunk 5 above)
Comment (epaulsen) [reply src/OpenTelemetry.Contrib.Instrumentation.Wcf/README.md]: Updated to a better code example which reflects usage of message.  
When enriching a WCF activity, you definitely want to have the message available.
@cijothomas Mind having another look?

File: src/OpenTelemetry.Contrib.Instrumentation.Wcf/README.md [hunk 7]
```diff
@@ -20,6 +20,33 @@ using var openTelemetry = Sdk.CreateTracerProviderBuilder()
...
+```csharp
+using var openTelemetry = Sdk.CreateTracerProviderBuilder()
+    .AddWcfInstrumentation(options =>
+    {
+        options.SuppressDownstreamInstrumentation = false;
+
+        // Enable enriching an activity after it is created.
+        options.Enrich = (activity, eventName,  message) =>
+        {            
+            var wcfMessage = message as Message;
```
Comment (cijothomas) [inline src/OpenTelemetry.Contrib.Instrumentation.Wcf/README.md:35]: this cast depends on the eventName ,right? If yes, please move it inside if(eventname=='') check.

File: test/OpenTelemetry.Contrib.Instrumentation.Wcf.Tests/TelemetryDispatchMessageInspectorTests.netfx.cs [hunk 8]
```diff
@@ -108,6 +110,22 @@ public void Dispose()
...
                     .AddWcfInstrumentation(options =>
                     {
+                        if (enrich)
+                        {
+                            options.Enrich = (activity, eventName, message) =>
+                            {
+                                switch (eventName)
+                                {
+                                    case WcfEnrichEventNames.AfterReceiveRequest:
+                                        activity.AddTag("server.afterreceiverequest", WcfEnrichEventNames.AfterReceiveRequest);
```
Comment (cijothomas) [inline test/OpenTelemetry.Contrib.Instrumentation.Wcf.Tests/TelemetryDispatchMessageInspectorTests.netfx.cs:120]:
```suggestion
                                        activity.SetTag("server.afterreceiverequest", WcfEnrichEventNames.AfterReceiveRequest);
```

File: test/OpenTelemetry.Contrib.Instrumentation.Wcf.Tests/TelemetryDispatchMessageInspectorTests.netfx.cs [hunk 9]
```diff
@@ -108,6 +110,22 @@ public void Dispose()
...
+                                        break;
+                                    case WcfEnrichEventNames.BeforeSendReply:
+                                        activity.AddTag("server.beforesendreply", WcfEnrichEventNames.BeforeSendReply);
```
Comment (cijothomas) [inline test/OpenTelemetry.Contrib.Instrumentation.Wcf.Tests/TelemetryDispatchMessageInspectorTests.netfx.cs:123]:
```suggestion
                                        activity.SetTag("server.beforesendreply", WcfEnrichEventNames.BeforeSendReply);
```

File: test/OpenTelemetry.Contrib.Instrumentation.Wcf.Tests/TelemetryClientMessageInspectorTests.cs [hunk 10]
```diff
@@ -147,6 +149,22 @@ public void Dispose()
...
                     .AddWcfInstrumentation(options =>
                     {
+                        if (enrich)
+                        {
+                            options.Enrich = (activity, eventName, message) =>
+                            {
+                                switch (eventName)
+                                {
+                                    case WcfEnrichEventNames.BeforeSendRequest:
+                                        activity.AddTag("client.beforesendrequest", WcfEnrichEventNames.BeforeSendRequest);
```
Comment (cijothomas) [inline test/OpenTelemetry.Contrib.Instrumentation.Wcf.Tests/TelemetryClientMessageInspectorTests.cs:159]: SetTag

File: test/OpenTelemetry.Contrib.Instrumentation.Wcf.Tests/TelemetryClientMessageInspectorTests.cs [hunk 11]
```diff
@@ -147,6 +149,22 @@ public void Dispose()
...
+                                        break;
+                                    case WcfEnrichEventNames.AfterReceiveReply:
+                                        activity.AddTag("client.afterreceivereply", WcfEnrichEventNames.AfterReceiveReply);
```
Comment (cijothomas) [inline test/OpenTelemetry.Contrib.Instrumentation.Wcf.Tests/TelemetryClientMessageInspectorTests.cs:162]: SetTag

File: src/OpenTelemetry.Contrib.Instrumentation.Wcf/README.md (hunk 7 above)
Comment (cijothomas) [reply src/OpenTelemetry.Contrib.Instrumentation.Wcf/README.md:35]: this cast always succeed, ok. Ignore my commend above.

**Review by cijothomas — APPROVED** [review]
couple of places uses AddTag. Please move to SetTag

**Comment by cijothomas** [comment]
> couple of places uses AddTag. Please move to SetTag

Can you fix the CI failures as well?

File: test/OpenTelemetry.Contrib.Instrumentation.Wcf.Tests/TelemetryClientMessageInspectorTests.cs (hunk 11 above)
Comment (epaulsen) [reply test/OpenTelemetry.Contrib.Instrumentation.Wcf.Tests/TelemetryClientMessageInspectorTests.cs]: Fixed ✔️

**Comment by epaulsen** [comment]
Fixed markdownlint errors and added more tests so that code coverage does not decrease.

**Comment by cijothomas** [comment]
@epaulsen One last CI failure remaining. Can you fix that too?

**Comment by epaulsen** [comment]
@cijothomas I think I fixed the last code check error, mind running the workflows again?  Sorry for all the failures, would have been easier if I could have executed the workflows my self.

**Comment by cijothomas** [comment]
> @cijothomas I think I fixed the last code check error, mind running the workflows again? Sorry for all the failures, would have been easier if I could have executed the workflows my self.

No problem. The workflows are automatically triggered. Since you are first-time contributor to this repo, it require a manual override by an approver. This shouldn't be an issue next time onwards!

**Comment by CodeBlanch** [comment]
https://www.nuget.org/packages/OpenTelemetry.Contrib.Instrumentation.Wcf/1.0.0-rc2

**Pull request merged**